### PR TITLE
Kyv onboard route

### PR DESCRIPTION
**Pull request by RingerJK**
## Description

Onboad navigation

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Please describe the PR goals. Just the stuff needed to implement the fix / feature and a simple rationale. It could contain many check points if needed

### Implementation

Please include all the relevant things implemented and also rationale, clarifications / disclaimers etc. related to the approach used. It could be as self code companion comments

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineTileVersions.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+
+/**
+ * This is a wrapper class for the [MapboxRouteTileVersions] class. It returns a list of
+ * all available versions of Routing Tiles available via [OfflineTiles]. This class
+ * encapsulates the unwrapping of the list from the response.
+ */
+internal class OfflineTileVersions {
```
Comment (Lebedinsky) [inline libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineTileVersions.kt]: Can we made it as `object`?

File: libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineRoute.kt [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            return Builder(onlineRouteBuilder)
+        }
+    }
+
+    /**
+     * Builds a URL string for offline.
+     *
+     * @return the offline url string
+     */
+    fun buildUrl(): String {
```
Comment (Guardiola31337) [inline libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineRoute.kt]: As discussed we should extract this out into an `interface` and use it across `libdirections-onboard` and `libdirections-offboard`.

File: libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineTileVersions.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * Call to receive all the available versions of Offline Tiles available.
+     *
+     * @param accessToken for the API call
+     * @param callback to be updated with the versions
+     */
+    fun fetchRouteTileVersions(accessToken: String, callback: OnTileVersionsFoundCallback) {
```
Comment (Guardiola31337) [inline libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineTileVersions.kt]: This is currently implemented but never used. As mentioned above, `MapboxOnboardRouter` is missing some APIs.

File: libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/task/ConfigureRouterTask.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package com.mapbox.navigation.route.onboard.task
+
+import android.os.AsyncTask
+import com.mapbox.navigation.route.onboard.OnOfflineTilesConfiguredCallback
+import com.mapbox.navigation.route.onboard.model.OfflineError
+import com.mapbox.navigator.Navigator
+
+internal class ConfigureRouterTask(
```
Comment (Guardiola31337) [inline libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/task/ConfigureRouterTask.kt]: This is currently implemented but never used. As mentioned above, `MapboxOnboardRouter` is missing some APIs.

File: libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/task/RemoveTilesTask.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package com.mapbox.navigation.route.onboard.task
+
+import android.os.AsyncTask
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.route.onboard.OnOfflineTilesRemovedCallback
+import com.mapbox.navigator.Navigator
+
+internal class RemoveTilesTask(
```
Comment (Guardiola31337) [inline libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/task/RemoveTilesTask.kt]: This is currently implemented but never used. As mentioned above, `MapboxOnboardRouter` is missing some APIs.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt [hunk 6]
```diff
@@ -1,9 +1,22 @@
...
 
 import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.route.model.Route
 
 interface Router {
 
-    fun getRoute(origin: Point, waypoints: List<Point>, callback: (route: Route) -> Unit)
+    fun getRoute(
+        origin: Point,
+        waypoints: List<Point>?,
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt]: Same as in https://github.com/mapbox/mapbox-navigation-android/pull/2240#discussion_r344938183

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/model/Route.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.mapbox.navigation.base.route.model
+
+import com.mapbox.api.directions.v5.models.RouteLeg
+
+data class Route(
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/route/model/Route.kt]: Same as in https://github.com/mapbox/mapbox-navigation-android/pull/2240#discussion_r344939985

Comment (Guardiola31337) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/model/Route.kt]: Same as in https://github.com/mapbox/mapbox-navigation-android/pull/2240#discussion_r344966253

File: libnavigation-util/src/main/java/com/mapbox/navigation/utils/time/Time.kt [hunk 8]
```diff
@@ -1,4 +1,4 @@
-package com.mapbox.services.android.navigation.v5.internal.utils
+package com.mapbox.navigation.util
```
Comment (Guardiola31337) [inline libnavigation-util/src/main/java/com/mapbox/navigation/utils/time/Time.kt]: Same as in https://github.com/mapbox/mapbox-navigation-android/pull/2240#discussion_r344967234

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 9]
```diff
@@ -21,6 +22,8 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
         navigator.updateLocation(rawLocation.toFixLocation())
     }
 
+    override fun getRoute(url: String): RouterResult = navigator.getRoute(url)
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt]: We're adding synchronization in the call-site 👀 https://github.com/mapbox/mapbox-navigation-android/blob/b1e01add2ce335a55a5ddd47790fc0483a5b00a1/libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/task/OfflineRouteRetrievalTask.kt#L36 Should we remove it from `OfflineRouteRetrievalTask` and add it here?

**Pull request closed without merging**